ui/src/file_upload.tsx

```diff
@@ -113,29 +113,42 @@ export const
       maxFileSizeBytes = max_file_size ? convertMegabytesToBytes(max_file_size) : 0,
       maxSizeBytes = max_size ? convertMegabytesToBytes(max_size) : 0,
       fileExtensions = file_extensions ? file_extensions.map(e => e.startsWith('.') ? e : `.${e}`) : null,
+      isAborted = React.useRef<B>(false),
+      xhrRef = React.useRef<XMLHttpRequest>(),
       upload = async (uploadFiles = files) => {
         const formData = new FormData()
         uploadFiles.forEach((f: File) => formData.append('files', f))
 
         try {
           const { responseText } = await new Promise<XMLHttpRequest>((resolve, reject) => {
-            const xhr = new XMLHttpRequest()
+            xhrRef.current = new XMLHttpRequest()
+            const xhr = xhrRef.current
             xhr.open("POST", wave.uploadURL)
             xhr.upload.onprogress = e => setPercentComplete(e.loaded / e.total)
             xhr.send(formData)
             xhr.onreadystatechange = () => {
               if (xhr.readyState !== XMLHttpRequest.DONE) return
               xhr.status >= 200 && xhr.status < 300 ? resolve(xhr) : reject(xhr)
             }
+            xhr.onabort = () => isAborted.current = true
           })
           const { files } = JSON.parse(responseText)
           wave.args[name] = files
 
           if (!compact) wave.push()
           setSuccessMsg(`Successfully uploaded files: ${files.map(({ name }: File) => name).join(',')}.`)
         }
-        catch ({ responseText }) { setError(responseText || 'There was an error when uploading file.') }
-        finally { setFiles([]) }
+        catch ({ responseText }) {
+          if (isAborted.current) {
+            isAborted.current = false
+            return
+          }
+          setError(responseText as S || 'There was an error when uploading file.')
+        }
```

Comment:
Should be cleaner and type-safe.
```suggestion
        catch (err: unknown) {
          let errMsg = 'There was an error when uploading file.'
          if (err instanceof XMLHttpRequest) {
            if (err.status === 0) return // Aborted.
            errMsg = err.responseText
          }
          setError(errMsg)
```